### PR TITLE
WIP: Adds Checksum validation solution for Drupal Views

### PR DESCRIPTION
**Pull request by stesi561**
Overview
----------------------------------------

Adds checksum access checks.

Before
----------------------------------------
No ability to grant access via checksum.

After
----------------------------------------
Adds a new Views Access Plugin that inherits from role - so we can grant access via checksum or being a logged on user.

image

Technical Details
----------------------------------------

Comments
----------------------------------------

To test create a basic view - doesn't even need to be of CiviCRM related content.
Add the access restriction CiviCRM checksum. 
A checksum can be generated via CiviCRM API 4 Explore Contact generateChecksum

Test that access is still granted via the role.
As an anon user test access is now granted via the checksum.
Add url params - cid1={contact_id}&cs={checksum}

I've followed the Webform pattern for now (with cid1) possibly this would be better with just cid?

Need some consideration of caching. I've hopefully added what is required such that this will not get cached - and as such we don't need to think about expiring when the checksum expires.

Probably needs some additional documentation

Release notes snippet
----------------------------------------

Adds a new views Access Plugin allowing you to grant access via CiviCRM checksum

### CODE REVIEW

File: src/Access/ContactChecksumCheckAccess.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+        $cid1 = filter_var(\Drupal::request()->query->get('cid1'), FILTER_VALIDATE_INT, ['options' => ['min_range' => 1]]);
+        $checksum =  \Drupal::request()->query->get('cs');
+
+        if (empty($cid1) || empty($checksum)) {
+          \Drupal::logger('ContactChecksumCheckAccess')->info('No cid1 or cs param');
+          return AccessResult::forbidden();
+        }
+
+        $civicrmAPI = \Drupal::service('civicrm_entity.api');
```
Comment (jackrabbithanna) [inline src/Access/ContactChecksumCheckAccess.php]: Can this be loaded via dependency injection? Also, could use the 'civicrm' service initialize() method

File: src/Plugin/views/access/ContactChecksum.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  public function access(AccountInterface $account) {
+    // Check if logged in and has access
+    $logged_in_access = parent::access($account);
+
+    if ($logged_in_access) {
+      return TRUE;
+    }
+
+    $cid1 = filter_var(\Drupal::request()->query->get('cid1'), FILTER_VALIDATE_INT, ['options' => ['min_range' => 1]]);
+    $checksum =  \Drupal::request()->query->get('cs');
```
Comment (jackrabbithanna) [inline src/Plugin/views/access/ContactChecksum.php]: If possible get the request object via dependency injection

File: src/Access/ContactChecksumCheckAccess.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+     * @return \Drupal\Core\Access\AccessResultInterface
+     *   The access result.
+     */
+    public function access(AccountInterface $account, Route $route) {
+        $options = unserialize($route->getRequirement('var_options'));
+        $account_roles = $account->getRoles();
+
+        $access_by_role = !empty(array_intersect(array_filter($options['role']), $account->getRoles()));
+        if ($access_by_role) {
+          \Drupal::logger('ContactChecksumCheckAccess')->info('Access by role');
```
Comment (jackrabbithanna) [inline src/Access/ContactChecksumCheckAccess.php]: If possible, use logger by dependency injection

**Comment by stesi561** [comment]
Thanks for the feedback. Agree with getting everything working using dependency injection is ideal.  Just wondering as well if this module is the correct place for this - was debating this verses a custom module or a PR on drupal civicrm module. It seems to fit in nicely with CiviCRM Entities views integration - however perhaps it makes sense that you would want to extend this to other places - i.e. have a content type that is "members only" but accessible via checksum so you can sent people smoothly to content via an email.

**Comment by jackrabbithanna** [comment]
I like the feature, and would put it in CE. It provides a good feature, does it right, and CE does have the central Views functionality for CiviCRM. It also provides an example if other devs want to do a similar custom thing. There's no worries wrt upgrade of the module. Can't think of a reason not to merge it. You're choice though, if you want to turn it into a separate contrib project, that's fine too.

**Comment by stesi561** [comment]
If I was to add some tests for this do you have any recommendations.

I assume we just need to do some functional JS to use browser to test that the access works?

**Comment by jackrabbithanna** [comment]
Yes, we have some tests that can be used as examples to start: https://github.com/eileenmcnaughton/civicrm_entity/blob/4.0.x/tests/src/FunctionalJavascript/Views/CivicrmAddressViewsTest.php

You'd want to create one or two contacts that could be listed in a view. 
You'd want to create and login with users, one without checksum which should get access denied, and acesss one with checksum that should show results. 

I'm inclined to merge this MR soon when you get ready. I'm working on an access control blog .. for many other facets of CiviCRM Entity, and this would be great to mention too. 

I still see some places that could use dependency injection, for the request object, logging.

File: civicrm_entity.services.yml [hunk 4]
```diff
@@ -25,3 +25,14 @@ services:
     arguments: ['@civicrm_entity.module_installer.inner', '%app.root%', '@module_handler', '@kernel', '@database', '@update.update_hook_registry']
     tags:
       - { name: service_collector, tag: 'module_install.uninstall_validator', call: addUninstallValidator }
+
+
+  civicrm_entity.contact_checksum_access_checker:
```
Comment (puresyntax71) [inline civicrm_entity.services.yml:30]: The dependencies are missing here.

File: src/Access/ContactChecksumCheckAccess.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    if (empty($cid1) || empty($checksum)) {
+      $this->getlogger('ContactChecksumCheckAccess')->info('No cid1 or cs param');
+      return AccessResult::forbidden();
+    }
+
+    $this->civicrmAPI = \Drupal::service('civicrm_entity.api');
+    $this->civicrmAPI->getFields('Contact');  // This forces a call to Civicrm initialize.
+
+    $results = \Contact::validateChecksum(FALSE)
```
Comment (puresyntax71) [inline src/Access/ContactChecksumCheckAccess.php:78]: Yeah, shouldn't this be just `Contact::validateChecksum()` i.e. without `\` at the beginning?

File: src/Access/ContactChecksumCheckAccess.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Logger\LoggerChannelTrait;
+use Drupal\civicrm_entity\CiviCrmApiInterface;
+
+use Symfony\Component\HttpFoundation\RequestStack;
+
+use Symfony\Component\Routing\Route;
+
+
```
Comment (puresyntax71) [inline src/Access/ContactChecksumCheckAccess.php:16]: Too many spaces here.

File: src/Plugin/views/access/ContactChecksum.php [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+use Drupal\Core\StringTranslation\TranslatableMarkup;
+use Drupal\user\RoleStorageInterface;
+
+use Drupal\civicrm_entity\CiviCrmApiInterface;
+
+use Symfony\Component\Routing\Route;
+
+use Drupal\user\Plugin\views\access\Role;
+use Symfony\Component\HttpFoundation\RequestStack;
+
```
Comment (puresyntax71) [inline src/Plugin/views/access/ContactChecksum.php:16]: Too many spaces here.

File: src/Access/ContactChecksumCheckAccess.php [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+   *   Run access checks for this account.
+   * @param \Symfony\Component\Routing\Route $route
+   *   The route for which an access check is being done.
+   *
+   * @return \Drupal\Core\Access\AccessResultInterface
+   *   The access result.
+   */
+  public function access(AccountInterface $account, Route $route) {
+    $options = unserialize($route->getRequirement('var_options'));
+    $account_roles = $account->getRoles();
```
Comment (puresyntax71) [inline src/Access/ContactChecksumCheckAccess.php:58]: Unused variable.

File: civicrm_entity.services.yml [hunk 9]
```diff
@@ -25,3 +25,14 @@ services:
...
+    class: Drupal\civicrm_entity\Access\ContactChecksumCheckAccess
+    autowire: true
+    tags:
+      - {name: access_check, applies_to: _civicrm_entity_checksum_access_check }
+
+  civicrm_entity.views_access_contactchecksum:
```
Comment (puresyntax71) [inline civicrm_entity.services.yml:36]: I don't think this is needed.

File: src/Plugin/views/access/ContactChecksum.php [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  /**
+   * Constructs a ContactChecksum object.
+   *
+   * @param \Symfony\Component\HttpFoundation\RequestStack $request_stack
+   *   The request stack.
+   * @param \Drupal\civicrm_entity\CiviCrmApiInterface $civicrm_api
+   *   The CiviCRM API bridge.
+   */
+  public function __construct(RequestStack $request_stack, CiviCrmApiInterface $civicrm_api) {
```
Comment (puresyntax71) [inline src/Plugin/views/access/ContactChecksum.php:56]: Missing the `public static function create()` implementation.

File: src/Plugin/views/access/ContactChecksum.php [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    if ($logged_in_access) {
+      return TRUE;
+    }
+    $request = $this->requestStack->getCurrentRequest();
+    $cid1 = filter_var($request->query->get('cid1'), FILTER_VALIDATE_INT, ['options' => ['min_range' => 1]]);
+    $checksum =  $request()->query->get('cs');
+
+    // Force CiviCRM to be intialized - ideally we'd use the api
+    // wrapper however it's api3 and we need api4.
+    $this->$civicrmAPI->getFields('Contact');  // This forces a call to Civicrm initialize.
```
Comment (puresyntax71) [inline src/Plugin/views/access/ContactChecksum.php:74]: Should be `$this->civicrmApi->getFIelds('Contact');`

File: src/Plugin/views/access/ContactChecksum.php [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+  public function access(AccountInterface $account) {
+    // Check if logged in and has access
+    $logged_in_access = parent::access($account);
+
+    if ($logged_in_access) {
+      return TRUE;
+    }
+    $request = $this->requestStack->getCurrentRequest();
+    $cid1 = filter_var($request->query->get('cid1'), FILTER_VALIDATE_INT, ['options' => ['min_range' => 1]]);
+    $checksum =  $request()->query->get('cs');
```
Comment (puresyntax71) [inline src/Plugin/views/access/ContactChecksum.php:70]: Should be `$requestion->query->get('cs');`.

File: src/Plugin/views/access/ContactChecksum.php [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+    // Force CiviCRM to be intialized - ideally we'd use the api
+    // wrapper however it's api3 and we need api4.
+    $this->$civicrmAPI->getFields('Contact');  // This forces a call to Civicrm initialize.
+
+    $results = \Contact::validateChecksum(FALSE)
```
Comment (puresyntax71) [inline src/Plugin/views/access/ContactChecksum.php:76]: Should just be `Contact::validateChecksum(FALSE)`.

File: civicrm_entity.services.yml (hunk 4 above)
Comment (mikeybeck) [reply civicrm_entity.services.yml:30]: Indeed - just tested and the following was required:

>     arguments:
>       - '@request_stack'
>       - '@civicrm_entity.api'

**Comment by puresyntax71** [comment]
Yeah, updated PR at #478.

**Comment by jackrabbithanna** [comment]
I merged #478 so closing this PR

**Pull request closed without merging**